lib/Dialect/HW/Transforms/HWSpecialize.cpp

```diff
@@ -267,50 +340,74 @@ void HWSpecializePass::runOnOperation() {
   ModuleOp module = getOperation();
 
   // Record unique module parameters and references to these.
-  llvm::DenseMap<hw::HWModuleOp, llvm::SetVector<ArrayAttr>>
-      uniqueModuleParameters;
   llvm::DenseMap<hw::HWModuleOp,
                  llvm::DenseMap<ArrayAttr, llvm::SmallVector<hw::InstanceOp>>>
       parametersUsers;
+  ParameterSpecializationRegistry registry;
 
   // Maintain a symbol cache for fast lookup during module specialization.
   SymbolCache sc;
   sc.addDefinitions(module);
 
   for (auto hwModule : module.getOps<hw::HWModuleOp>()) {
+    // If this module is parametric, defer registering its parametric
+    // instantiations until this module is specialized
+    if (!hwModule.parameters().empty())
+      continue;
     for (auto instanceOp : hwModule.getOps<hw::InstanceOp>()) {
-      auto *targetOp = sc.getDefinition(instanceOp.moduleNameAttr());
-      auto targetHWModule = dyn_cast<hw::HWModuleOp>(targetOp);
-      if (!targetHWModule)
-        continue; // Won't specialize external modules.
+      if (auto targetHWModule = targetModuleOp(instanceOp, sc)) {
+        auto parameters = instanceOp.parameters();
+        registry.registerModuleOp(targetHWModule, parameters, sc);
 
-      if (targetHWModule.parameters().size() == 0)
-        continue; // nothing to record or specializeauto paramValue =.
-
-      auto parameters = instanceOp.parameters();
-      uniqueModuleParameters[targetHWModule].insert(parameters);
-      parametersUsers[targetHWModule][parameters].push_back(instanceOp);
+        parametersUsers[targetHWModule][parameters].push_back(instanceOp);
+      }
     }
   }
 
   // Create specialized modules.
   OpBuilder builder = OpBuilder(&getContext());
   builder.setInsertionPointToStart(module.getBody());
-  for (auto it : uniqueModuleParameters) {
-    for (auto parameters : it.getSecond()) {
-      HWModuleOp specializedModule;
-      if (failed(specializeModule(builder, parameters, sc, it.getFirst(),
-                                  specializedModule))) {
-        signalPassFailure();
-        return;
+  llvm::DenseMap<hw::HWModuleOp, llvm::DenseMap<ArrayAttr, hw::HWModuleOp>>
+      specializations;
+
+  while (true) {
```

Comment:
I think the combination of the `while (true), registry, nextRegistry` deserves a small comment to save a minute or two of lulling over what this code does.

Reply:
I added a comment. I also changed comments referring to "specialization passes" to "specialization loops" to disambiguate the looping from the pass as a whole.